add Zip/Repeat typeclass with instances (#1814)
arrow-kt/arrow: modules/docs/arrow-docs/docs/docs/arrow/typeclasses/zip/README.md

---
layout: docs
title: Zip
permalink: /docs/arrow/typeclasses/zip/
redirect_from:
  - /docs/typeclasses/zip/
---

## Zip

{:.beginner}
beginner

> The `Zip` typeclass extends the `Semialign` typeclass with an function that takes the intersection of non-uniform shapes.

Reviewer: ```suggestion
The `Zip` typeclass extends the `Semialign` typeclass with a function that takes the intersection of non-uniform shapes.
```

### Main Combinators

#### zip

> Combines to structures by taking the intersection of their shapes 

Reviewer: ```suggestion
Combines two structures by taking the intersection of their shapes 
```

and using `Tuple2` to hold the elements.

`fun <A, B> Kind<F, A>.zip(other: Kind<F, B>): Kind<F, Tuple2<A, B>>`

```kotlin:ank
import arrow.core.extensions.*
import arrow.core.extensions.listk.zip.zip
import arrow.core.*

ListK.zip().run {
    listOf("A", "B").k().zip(listOf(1, 2, 3).k())
}
```

#### zipWith

> Combines to structures by taking the intersection of their shapes

Reviewer: ```suggestion
Combines two structures by taking the intersection of their shapes
```

> and combining the elements with the given function.

Reviewer: ```suggestion
and then combines the elements with the given function.
```

`fun <A, B, C> Kind<F, A>.zipWith(other: Kind<F, B>, f: (A, B) -> C): Kind<F, C>`

```kotlin:ank
import arrow.core.extensions.*
import arrow.core.extensions.listk.zip.zip
import arrow.core.*

ListK.zip().run {
    listOf("A", "B").k().zipWith(listOf(1, 2, 3).k()) {
        a, b -> "$a # $b"
    }
}
```

### Laws

Arrow provides [`ZipLaws`][functor_laws_source]{:target="_blank"} in the form of test cases for internal verification of lawful instances and third party apps creating their own Zip instances.

#### Creating your own `Zip` instances

> Arrow already provides Align instances for common datatypes (e.g. Option, ListK, MapK). See their implementations

Reviewer: ```suggestion
Arrow already provides Zip instances for common datatypes (e.g. Option, ListK, MapK). See their implementations
```
